In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sea

In [40]:
train = pd.read_csv("D:/Kaggle/Housing_Prices/train.csv")
print(train.shape)
test = pd.read_csv("D:/Kaggle/Housing_Prices/test.csv")
SalePrice = train['SalePrice']
train = train.drop(columns = ['SalePrice'])
df = pd.concat([train, test])
index = df['Id']


(1460, 81)
(1460, 80)
(1459, 80)
(2919, 80)


In [43]:
print(train.shape)
print(test.shape)
print(df.shape)


(1460, 80)
(1459, 80)
(2919, 80)
0       208500
1       181500
2       223500
3       140000
4       250000
         ...  
1455    175000
1456    210000
1457    266500
1458    142125
1459    147500
Name: SalePrice, Length: 1460, dtype: int64


In [37]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

['MSZoning', 'LotFrontage', 'Alley', 'Utilities', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Electrical', 'BsmtFullBath', 'BsmtHalfBath', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType']


In [38]:
df[Cols_Null].describe()

LotFrontage   MasVnrArea   BsmtFinSF1   BsmtFinSF2    BsmtUnfSF  \
count  2433.000000  2896.000000  2918.000000  2918.000000  2918.000000   
mean     69.305795   102.201312   441.423235    49.582248   560.772104   
std      23.344905   179.334253   455.610826   169.205611   439.543659   
min      21.000000     0.000000     0.000000     0.000000     0.000000   
25%      59.000000     0.000000     0.000000     0.000000   220.000000   
50%      68.000000     0.000000   368.500000     0.000000   467.000000   
75%      80.000000   164.000000   733.000000     0.000000   805.500000   
max     313.000000  1600.000000  5644.000000  1526.000000  2336.000000   

       TotalBsmtSF  BsmtFullBath  BsmtHalfBath  GarageYrBlt   GarageCars  \
count  2918.000000   2917.000000   2917.000000  2760.000000  2918.000000   
mean   1051.777587      0.429894      0.061364  1978.113406     1.766621   
std     440.766258      0.524736      0.245687    25.574285     0.761624   
min       0.000000      0.000000      0.000000  1895.000000     0.000000   
25%     793.000000      0.000000      0.000000  1960.000000     1.000000   
50%     989.500000      0.000000      0.000000  1979.000000     2.000000   
75%    1302.000000      1.000000      0.000000  2002.000000     2.000000   
max    6110.000000      3.000000      2.000000  2207.000000     5.000000   

        GarageArea  
count  2918.000000  
mean    472.874572  
std     215.394815  
min       0.000000  
25%     320.000000  
50%     480.000000  
75%     576.000000  
max    1488.000000

In [39]:
df[Cols_Null].isnull().sum()

MSZoning           4
LotFrontage      486
Alley           2721
Utilities          2
Exterior1st        1
Exterior2nd        1
MasVnrType        24
MasVnrArea        23
BsmtQual          81
BsmtCond          82
BsmtExposure      82
BsmtFinType1      79
BsmtFinSF1         1
BsmtFinType2      80
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
Electrical         1
BsmtFullBath       2
BsmtHalfBath       2
KitchenQual        1
Functional         2
FireplaceQu     1420
GarageType       157
GarageYrBlt      159
GarageFinish     159
GarageCars         1
GarageArea         1
GarageQual       159
GarageCond       159
PoolQC          2909
Fence           2348
MiscFeature     2814
SaleType           1
dtype: int64

For variables where NA just means the item doesn't exist we will update to none. 

Check PoolQC NA against PoolArea since very little properties have a pool.

We will remove Alley and MiscFeature since a large portion of data is null. 

In [7]:
df['PoolArea'].describe()

count    1460.000000
mean        2.758904
std        40.177307
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max       738.000000
Name: PoolArea, dtype: float64

remove Pool Area in addition to PoolQC, provides little value since out of 1460 obs only 7 have a pool. 

In [8]:
df[Cols_Null].dtypes

LotFrontage     float64
Alley            object
MasVnrType       object
MasVnrArea      float64
BsmtQual         object
BsmtCond         object
BsmtExposure     object
BsmtFinType1     object
BsmtFinType2     object
Electrical       object
FireplaceQu      object
GarageType       object
GarageYrBlt     float64
GarageFinish     object
GarageQual       object
GarageCond       object
PoolQC           object
Fence            object
MiscFeature      object
dtype: object

### Update NA in LotFrontage with the mean of LotFrontages with similar LotAreas. 

In [9]:
df['LotArea'].describe()

count      1460.000000
mean      10516.828082
std        9981.264932
min        1300.000000
25%        7553.500000
50%        9478.500000
75%       11601.500000
max      215245.000000
Name: LotArea, dtype: float64

In [10]:
df["LotFrontage"].fillna(df['LotFrontage'].mean(), inplace = True) 
df['LotFrontage'].mean()

70.04995836802642

In [11]:
df.drop(['Alley'], axis=1, inplace = True)
df.drop(['PoolQC'], axis=1, inplace = True)
df.drop(['PoolArea'], axis=1, inplace = True)
df.drop(['Fence'], axis=1, inplace = True)
df.drop(['MiscFeature'], axis=1, inplace = True)

In [12]:
df.isnull().sum()


Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
SalePrice        0
Length: 76, dtype: int64

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 76 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1460 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non

In [14]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

['MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond']


In [15]:
df['MasVnrType'].describe

<bound method NDFrame.describe of 0       BrkFace
1          None
2       BrkFace
3          None
4       BrkFace
         ...   
1455       None
1456      Stone
1457       None
1458       None
1459       None
Name: MasVnrType, Length: 1460, dtype: object>

In [16]:
df['MasVnrType'].fillna('None', inplace = True) 
df['MasVnrArea'].fillna('None', inplace = True) 
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond']


In [17]:
df['GarageType'].fillna('No Garage', inplace = True) 
df['GarageYrBlt'].fillna('No Garage', inplace = True) 
df['GarageFinish'].fillna('No Garage', inplace = True) 
df['GarageQual'].fillna('No Garage', inplace = True) 
df['GarageCond'].fillna('No Garage', inplace = True) 
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)

['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu']


In [18]:
df['FireplaceQu'].fillna('No Fireplace', inplace = True)

In [19]:
df['Electrical'].fillna(df['Electrical'].mode()[0], inplace = True)
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)




['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']


Replace BSMT Bathrooms with No Basement if Basement Qual is NA

In [20]:
df.loc[df['BsmtQual'].isnull() & df['BsmtFullBath'].eq(0), 'BsmtFullBath'] = 'No Basement'
df.loc[df['BsmtQual'].isnull() & df['BsmtHalfBath'].eq(0), 'BsmtHalfBath'] = 'No Basement'
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)


['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']


In [21]:
df[Cols_Null].isnull().sum()

BsmtQual        37
BsmtCond        37
BsmtExposure    38
BsmtFinType1    37
BsmtFinType2    38
dtype: int64

In [22]:
df['BsmtFullBath'] = df['BsmtFullBath'].apply(str)
print(np.unique(df[['BsmtFullBath']]))

['0' '1' '2' '3' 'No Basement']


In [23]:
df['BsmtHalfBath'] = df['BsmtHalfBath'].apply(str)
print(np.unique(df[['BsmtHalfBath']]))

['0' '1' '2' 'No Basement']


In [25]:
df['BsmtQual'].fillna('No Basement', inplace = True) 
df['BsmtCond'].fillna('No Basement', inplace = True) 
df['BsmtExposure'].fillna('No Basement', inplace = True) 
df['BsmtFinType1'].fillna('No Basement', inplace = True) 
df['BsmtFinType2'].fillna('No Basement', inplace = True) 

In [28]:
Cols_Null = df.columns[df.isnull().any()].tolist()
print(Cols_Null)
df.info()


[]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 76 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1460 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 

Update categorical columns as strings and creat dummy variables